manta-pay/src/crypto/hash/poseidon_parameter_generation/lfsr.rs

```diff
@@ -0,0 +1,210 @@
+
+            // Obtain the second bit
+            res.push(self.update());
+        }
+
+        res
+    }
+
+    pub fn get_field_elements_rejection_sampling<S, COM>(
+        &mut self,
+        num_elems: usize,
+    ) -> Vec<S::ParameterField> 
+    where S: Specification<COM>
+    {
+        assert_eq!(S::MODULUS_BITS as u64, self.prime_num_bits);
+
+        let mut res = Vec::new();
+        for _ in 0..num_elems {
+            // Perform rejection sampling
+            loop {
+                // Obtain n bits and make it most-significant-bit first
+                let mut bits = self.get_bits(self.prime_num_bits as usize);
+                bits.reverse();
+
+                // Construct the number
+                if let Some(f) = S::try_from_bits_le(&bits) {
+                    res.push(f);
+                    break;
+                }
+            }
+        }
+
+        res
+    }
+
+    pub fn get_field_elements_mod_p<S, COM>(&mut self, num_elems: usize) -> Vec<S::ParameterField> 
+    where S: Specification<COM>
+    {
+        assert_eq!(S::MODULUS_BITS as u64, self.prime_num_bits);
+
+        let mut res = Vec::new();
+        for _ in 0..num_elems {
+            // Obtain n bits and make it most-significant-bit first
+            let mut bits = self.get_bits(self.prime_num_bits as usize);
+            bits.reverse();
+
+            let bytes = bits
+                .chunks(8)
+                .map(|chunk| {
+                    let mut result = 0u8;
+                    for (i, bit) in chunk.iter().enumerate() {
+                        result |= u8::from(*bit) << i
+                    }
+                    result
+                })
+                .collect::<Vec<u8>>();
+
+            res.push(S::from_le_bytes_mod_order(&bytes));
+        }
+
+        res
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::GrainLFSR;
+    use crate::crypto::constraint::arkworks::{Fp, R1CS};
+    use crate::crypto::hash::poseidon::arkworks::Specification;
+    use ark_ff::field_new;
+
+    /// Compiler Type
+    type Compiler<S> = R1CS<<S as Specification>::Field>;
+    pub type ConstraintField = ark_bls12_381::Fr;
+
+    #[derive(Clone)]
+    pub struct PoseidonSpec;
+
+    // Only for test purpose
+    impl Specification for PoseidonSpec {
+        type Field = ConstraintField;
+        const FULL_ROUNDS: usize = 8;
+        const PARTIAL_ROUNDS: usize = 55;
+        const SBOX_EXPONENT: u64 = 5;
+    }
+
+    #[test]
+    fn test_grain_lfsr_consistency() {
+        // sage generate_parameters_grain_deterministic.sage 1 0 255 3 8 55 0x73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001
+
+        let mut lfsr = GrainLFSR::new(255, 3, 8, 55);
+        assert_eq!(
+            lfsr.get_field_elements_rejection_sampling::<PoseidonSpec, Compiler<PoseidonSpec>>(1)[0],
+            Fp(field_new!(
+                ConstraintField,
+                "41764196652518280402801918994067134807238124178723763855975902025540297174931"

```

Comment:
We can add the sage script in plonk-prototype to show how to generate the test case for this hardcode test. Those scripts should be correct (or I messed up)

Reply:
I created an issue #67 for this